fastdtw/_fastdtw.pyx

```diff
@@ -0,0 +1,166 @@
+
+from __future__ import absolute_import, division
+
+from libcpp.vector cimport vector  # noqa
+from libcpp.map cimport map
+from libcpp.set cimport set
+from libcpp.utility cimport pair
+from cython.operator cimport dereference as deref
+from cython.operator cimport preincrement
+
+cdef struct CostEntry:
+    double cost
+    int prev_x_idx
+    int prev_y_idx
+
+
+def fastdtw(x, y, radius=1, dist=lambda a, b: abs(a - b)):
+    min_time_size = radius + 2
+
+    if len(x) < min_time_size or len(y) < min_time_size:
+        return dtw(x, y, dist=dist)
+
+    x_shrinked = __reduce_by_half(x)
+    y_shrinked = __reduce_by_half(y)
+    distance, path = fastdtw(x_shrinked, y_shrinked, radius=radius, dist=dist)
+    window = __expand_window(path, len(x), len(y), radius)
+    return dtw(x, y, window, dist=dist)
+
+
+def dtw(x, y, window=None, dist=lambda a, b: abs(a - b)):
+    cdef unsigned int len_x, len_y, i, j, k, len_window
+
+    cdef double inf, cost_up, cost_left, cost_cnr, dt
+    cdef pair[int, int] key_up, key_left, key_cnr, key_next
+    cdef vector[pair[int, int]] window_vec
+    len_x, len_y = len(x), len(y)
+    if window is None:
+        window = [(i, j) for i in range(len_x) for j in range(len_y)]
+
+    len_window = len(window)
+    window_vec.resize(len_window)
+    for i in range(len_window):
+        window_vec[i].first = window[i][0] + 1
+        window_vec[i].second += window[i][1] + 1
+
+    cdef map[pair[int, int], CostEntry] D
+    cdef CostEntry cost_entry
+
+    cost_entry.cost = 0
+    cost_entry.prev_x_idx = 0
+    cost_entry.prev_y_idx = 0
+    inf = float('inf')
+
+    D[__get_key(0, 0)] = cost_entry
+
+    for k in range(len_window):
+        i = window_vec[k].first
+        j = window_vec[k].second
+
+        dt = dist(x[i-1], y[j-1])
+
+        cost_up = __get_cost(D, i - 1, j, inf)
+        cost_left = __get_cost(D, i, j - 1, inf)
+        cost_cnr = __get_cost(D, i - 1, j - 1, inf)
+
+        if cost_up < cost_left and cost_up < cost_cnr:
+            cost_entry.cost = cost_up + dt
+            cost_entry.prev_x_idx = i - 1
+            cost_entry.prev_y_idx = j
+
+        elif cost_left < cost_cnr:
+            cost_entry.cost = cost_left + dt
+            cost_entry.prev_x_idx = i
+            cost_entry.prev_y_idx = j - 1
+
+        else:
+            cost_entry.cost = cost_cnr + dt
+            cost_entry.prev_x_idx = i - 1
+            cost_entry.prev_y_idx = j - 1
+
+        D[__get_key(i, j)] = cost_entry
+
+    path = []
+    i, j = len_x, len_y
+    while not (i == j == 0):
+        path.append((i-1, j-1))
+        cost_entry = D[__get_key(i, j)]
+        i = cost_entry.prev_x_idx
+        j = cost_entry.prev_y_idx
+    path.reverse()
+    return (D[len_x, len_y].cost, path)
+
+
+cdef __reduce_by_half(x):
+    cdef int i, len_x = len(x)
+    out = []
+    for i in range(0, len_x - len_x % 2, 2):
+        out.append((x[i] + x[1+i]) / 2.0)
+    return out
```

Comment:
I don't think this is much better than list comprehension, if at all. If numpy is an allowed dependency of the project (sensible sure is: @slaypni can confirm), then something like:
```cython
cdef inline double[:] __reduce_by_half(x):
    cdef:
        int i, j
        double[:] reduced = np.empty(x.shape[0] // 2)
    with nogil:
        for j, i in enumerate(range(0, x.shape[0] - x.shape[0] % 2, 2)):
            reduced[j] = (x[i] + x[i + 1]) / 2.
    return reduced
```
would be better. This returns a refcounted (memory-managed) [typed memoryview](https://webcache.googleusercontent.com/search?q=cache:HYPhFTJPolMJ:docs.cython.org/src/userguide/memoryviews.html+&cd=1&hl=en&ct=clnk&gl=us), which is super-fast to manipulate.